src/core/mutatus.scala

```diff
@@ -257,29 +272,39 @@ case class LongId(id: Long) extends IdKey {
 }
 
 /** a data access object for a particular type */
-case class Dao[T](kind: String)(implicit svc: Service, namespace: Namespace) {
+case class Dao[T](kind: String)(
+    implicit svc: Service,
+    namespace: Namespace,
+    decoder: Decoder[T]
+) {
 
   private[mutatus] lazy val keyFactory = {
     val baseFactory = svc.readWrite.newKeyFactory().setKind(kind)
     namespace.option.foldLeft(baseFactory)(_.setNamespace(_))
   }
 
-  /** returns an iterator of all the values of this type stored in the GCP Platform */
-  def all()(implicit decoder: Decoder[T]): Iterator[T] =
-    QueryBuilder(kind).find()(svc, namespace, decoder)
-
-  /** returns query builder */
-  def query(implicit decoder: Decoder[T]): QueryBuilder[T] =
-    QueryBuilder[T](kind)
+  /** returns query builder with empty criteria which fetches all the values of this type stored in the GCP Platform */
+  def all: QueryBuilder[T] = QueryBuilder[T](kind)
+  def query: QueryBuilder[T] = all
 
-  def unapply[R](id: R)(implicit decoder: Decoder[T], idField: IdField[T] {
+  def unapply[R](id: R)(implicit idField: IdField[T] {
     type Return = R
   }): Option[T] = {
     val key = idField.idKey(id).newKey(keyFactory)
     Try(decoder.decode(svc.read.get(key))).toOption
   }
 }
 
+/** companion object for data access objects */
+object Dao {
+  implicit def apply[T](
```

Comment:
I see why you've done it, though I'm a little uneasy about an implicit method that's also an `apply` method (and which also takes its own implicits). I'm not sure it's wrong...

Reply:
I was independently wondering whether we could provide the `Dao` functionality through an extension method on companion objects. But I think this can go into a separate issue/PR.

Reply:
In this case I've only added param `decoder:Decoder[T]` and moved companion object closer to case class definition (Before it was at the end of file, which made it bit annoying). 
I don't think it is wrong, although I don't remember seeing such such construct. I don't know what type of problems it may cause, but in tests it worked out well.

Reply:
Ok, so it's something I wrote! I probably had this argument with myself at the time, and then forgot about it. :) No need to change anything for this PR.